include/seqan/index/index_bifm.h

```diff
@@ -0,0 +1,334 @@
+ *
+ * These tags can be used to get @link Fibre Fibres @endlink of a Bidirectional FM index.
+ *
+ * @see Fibre
+ * @see Index#getFibre
+ *
+ * @tag BidirectionalFMIndexFibres#FibreText
+ * @brief The original text of the index.
+ *
+ * @tag BidirectionalFMIndexFibres#FibreSA
+ * @brief The compressed suffix array of the text.
+ *
+ * @tag BidirectionalFMIndexFibres#FibreLF
+ * @brief The lf table.
+ */
+
+template <typename TText, typename TSpec, typename TConfig>
+struct Fibre<Index<TText, BidirectionalFMIndex<TSpec, TConfig> >, FibreLF>
+{
+    typedef LF<TText, TSpec, TConfig>  Type;
+};
+
+template <typename TText, typename TSpec, typename TConfig>
+struct Fibre<Index<TText, BidirectionalFMIndex<TSpec, TConfig> >, FibreSA>
+{
+    typedef CompressedSA<TText, TSpec, TConfig> Type;
+};
+
+template <typename TText, typename TSpec, typename TConfig>
+struct Fibre<Index<TText, BidirectionalFMIndex<TSpec, TConfig> >, FibreTempSA>
+{
+    typedef Index<TText, BidirectionalFMIndex<TSpec, TConfig> >          TIndex_;
+    typedef typename SAValue<TIndex_>::Type                 TSAValue_;
+
+    // NOTE(esiragusa): External causes problems on device code.
+#ifndef PLATFORM_CUDA
+    typedef String<TSAValue_, External<ExternalConfigLarge<> > >                Type;
+#else
+    typedef String<TSAValue_, typename DefaultIndexStringSpec<TText>::Type>     Type;
+#endif
+};
+
+
+
+// ----------------------------------------------------------------------------
+// Class BidirectionalFMIndex
+// ----------------------------------------------------------------------------
+
+/*!
+ * @class BidirectionalFMIndex
+ * @extends Index
+ * @headerfile <seqan/index.h>
+ * @brief A bidirectional index based on the Burrows-Wheeler transform.
+ *
+ * @signature template <typename TText[, typename TSpec[, typename TConfig]]>
+ *            class Index<TText, BidirectionalFMIndex<TSpec, TConfig> >;
+ *
+ * @tparam TText   The text type. Types: @link String @endlink, @link StringSet @endlink
+ * @tparam TSpec   Bidirectional FM index specialisation, defaults to <tt>void</tt>.
+ * @tparam TConfig A config object which determines the data types of the different fibres, defaults to
+ *                 <tt>FMIndexConfig&lt;TSpec&gt;</tt>. TBidirectional is only used internally to
+ *                 distinguish between a stand-alone FM index and an FM index belonging to a pair of
+ *                 FM indices forming a bidirectional FM index.
+ *
+ * @section Structure
+ *
+ * The Bidirectional FM index consists of two FM indices that index the original text and the reverse text.
+ * Performing a regular backward search on one on the FM indices will automatically update the search
+ * interval on the other index.
+ */
+
+template <typename TText, typename TSpec, typename TSpec2, typename TLengthSum, typename TBidirectional>
+class Index<TText, BidirectionalFMIndex<TSpec, FMIndexConfig<TSpec2, TLengthSum, TBidirectional> > >
```

Comment:
The class signature should be
```class Index<TText, BidirectionalIndex<TSpec> >```
to allow any bidirectional index in the future.